pkg/ring/batch.go

```diff
@@ -115,11 +124,11 @@ func DoBatch(ctx context.Context, op Operation, r ReadRing, keys []uint32, callb
 
 	wg.Add(len(instances))
 	for _, i := range instances {
-		go func(i instance) {
+		e.Submit(func() {
 			err := callback(i.desc, i.indexes)
 			tracker.record(i, err)
 			wg.Done()
-		}(i)
+		})
```

Comment:
The closure var  should be fine now:

https://tip.golang.org/doc/go1.22

>Previously, the variables declared by a “for” loop were created once and updated by each iteration. In Go 1.22, each iteration of the loop creates new variables, to avoid accidental sharing bugs. The transition support tooling described in the proposal continues to work in the same way it did in Go 1.21.